torch/_dynamo/guards.py

```diff
@@ -413,6 +413,10 @@ def GRAD_MODE(self, guard: Guard):
             code = "not ___is_grad_enabled()"
         self._produce_guard_code(guard, [code])
 
+    @staticmethod
+    def SYMBOL_MATCH():
+        pass
```

Comment:
You still want to merge this into master? Looks sus

Reply:
+1

Reply:
Yes, I do, this is for symbolic shape guard comparison.

Reply:
Let me add a doc

Reply:
Looking at the use case, it looks like you're manually attaching code to the `SYMBOL_MATCH` guard which is the actual code to be run? @jansel does this make sense to you?

Reply:
How exactly does the manual attaching code work?  I don't fully understand this.

Reply:
Hey, sure thing.

This is just a little hack to get things going. Its not amazing. 

Story time.

When we call torch._dynamo.export(), we get back a `graph` and `guards`. The graph is the whole capture graph, the `guards` is a structure containing all the guards we made.

Most guards follow a really clear lifecycle:

- Accumulate as we trace
- Install by writing into check_fn
- Associate with the frame
- Check on subsequent frame

You know the drill. For symbolic shapes, the lifecycle is identical, EXCEPT the implementation mechanism is not.

We accumulate it via shape_env, not via funcs stored in guards.py. 

This is fine, works great, and doesn't use `SYMBOL_MATCH` AT ALL. So why do we add this function?

Each guard has an associated creation function that we can use to determine what kind of guard this was. TYPE_CHECK, TENSOR_CHECK, etc etc. Symbolic shape guards did not use the same accumulation paths, and so do not inherently have one of these. Normally, no one uses or cares about this create_fn tracking, its just a little tag unless you are exporting (Export cares about it as part of its contract). This just creates an empty func we can use for tagging. You could call it anything, and put it anywhere.

Reply:
In that case, it seems like it would be clearer to just say something like SYMBOL_MATCH = object() (to prevent people from calling it) and also document that the guard creation functions are being used as tags for export. Though this seems like an unintended use of the system, and I'm curious if @jansel had a prior for how to represent tagged guards (tbh, the whole guards architecture doesn't make much sense to me, haven't spend enough time thinking about it)

Reply:
Sure, open to doing anything else. `SYMBOL_MATCH = object()` is fine. Do you want me to find an hour to go over guards in more detail?